### PR TITLE
RTCIceCandidate and RTCDataChannel

### PR DESCRIPTION
**Pull request by nvanofwegen**
I used your repo as a baseline to rebuild the work I've done on top of gt2k's work. Currently the data channels work again between editor/standalone. But browser-browser and editor/standalone - browser doesn't work anymore. The channels do open, but no data is received at the other end. However, I'm using negotiated = true in the options, which I think it not the way it should be done in the end. I'm also only able to test and verify on windows/webgl.

This PR has the following additions/fixes
- Latest com.unity.webrtc commits
- Added TextMessages on RTCDataChannels
- Renamed VideoStreamTrack.jslib file
- Removed WebSocketSignaling (not part of WebRTC, should be in the project folder)
- RTCDataChannel options using JsonUtility (crash on newtonsoft in the browser)
- Added missing jslib calls (Context/RtpReceiver)
- Removed the audio/video offer offer/answer options
- Fixes on RTCIceCandidates

Question/suggestion:
- I'm using this repo as a submodule in the folder Assets/Packages/com.unity.webrtc instead of Packages/com.unity.webrtc, because else the changes are not used while building. Did you manage to fix this?
- Since we can make PR between forks (I didn't know this was possible) I suggest we keep PR'ing between the forks with everyone that wants to work on this until something rolls out.

### CODE REVIEW

**Comment by antoined73** [comment]
Thank you very much for your help !

I personnaly use this repository as a normal package in the editor at Packages/com.unity.webrtc
I haven't encountered the problem you're describing. Are you talking about changes in the WebGL plugin (.jslib files) ?

Yes, let's keep doing PRs until something rolls out.

**Comment by nvanofwegen** [comment]
Hm no, I was under the impression it happened and then I thought of the comment in the [WebGL issue](https://github.com/Unity-Technologies/com.unity.webrtc/issues/349#issuecomment-838024524). But it's not that big a deal, I just need to watch I don't commit the Lib path. I'm not sure anymore if it was the jslib of c# code that didn't get updated.

Currently I'm busy with getting the datachannels to work just within the editor. I encountered a problem that the RTCDataChannels do connect, but disconnects after ~2/3 seconds. When doing the same approach/sequence just with pure javascript, it stays connected. So maybe there is a bug in the unity code itself.

**Comment by antoined73** [comment]
Yes it's clear that gtk2k already talked about that issue, but I didn't encountered it yet so...

Ok great !
On my side, I'm pausing devlopment on it for now, I have greater priorities to handle at work for now.
Did you manage to run a sample scene in a webgl build yet ?
I couldn't make it work in the scene MultiVideoReceive, I get this error when I launch the call : 
` Uncaught ReferenceError: _CreatePeerConnectionWithConfig is not defined at _ContextCreatePeerConnectionWithConfig`

Which is weird because the method CreatePeerConnectionWithConfig effectively exist in RTCPeerConnection.jslib.

PS : I'll check by moving the Package to Assets, maybe it is indeed a problem about built code not using the last package version...

**Comment by antoined73** [comment]
Update : 
So yeah in fact it seems that moving the files to the Assets resolves the previous error I mentionned (_CreatePeerConnectionWithConfig_). So I'll stuck with devlopping the packages in Assets for now.

Now the MultiVideoReceive scene can launch a call, add video objetcs.
But it is throwing the error "cannot read video of undefined" when adding tracks.
It seems to be at this line in VideoRenderer.jslib, in UpdateRendererTexture method : 
`var video = uwcom_remoteVideoTracks[trackPtr].video;`
Where uwcom_remoteVideoTracks[trackPtr] is undefined.

https://i.ibb.co/d6zHZcJ/Web-GL-error.png

Seems that trackPtr is wrong.

**Comment by nvanofwegen** [comment]
Alright!

Yesterday I managed to get the DataChannels working, it's seems that for each PeerConnection you need two DataChannels, a receiving and sending one, and the onmessage listener was on the sender.
There are still some issues with the negotiation. Between standalone/editor and webgl it works all the time, but when connecting between browsers it works 'sometimes'. So I think it due to some async/await stuff that needs to be included, but I'll have a look at [perfect negotiation](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Perfect_negotiation) first for that.

Next on my list would be investigating audio/video using [getUserMedia](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia) and get the data into the MediaStream(Track), as it seems that currently it only it aimed for grabbing the Audio and Camera of unity itself (to stream your game).

Some tip if you didn't know it yet: Since webgl building takes ages, all the jslib code is in the *.framework.js file, so to develop/test I change that file and when a found a final working solution I copy it in the actual .jslib files. Saves a lot of time!

**Comment by nvanofwegen** [comment]
Update!

TL;DR:
- Sending audio/video using getUserMedia works
- Receiving video in unity works, audio does not (both video/audio are received in the DOM tho!)
- RTCDataChannels creation still unstable

I looked into the stability of the creation of the RTCDataChannels and can't seem to find anything strange for now, but that's mainly because of my knowledge of WebRTC. 
However, I added getUserMedia to the MediaStream and (for now, in a hacky way) injected the resulted tracks into an already existing MediaStream object inside unity (I couldn't find out how to quickly create a MediaStream object in unity from the jslib code, since it's async stuff. Also the registration of the VideoStreamTrack and manage objects pointers in jslib is not something I'm happt with). But on the other end (another browser) the data was received and added to a HTMLVideoElement and correctly binding to a texture in Unity. The HTMLVideoElement in the background does receive the audio, so the actual transfer is working. Luckily, handling audio inside of unity seems to be around the corner https://github.com/Unity-Technologies/com.unity.webrtc/pull/440

Back to the RTCDataChannels: I create two channels (reliable/unreliable) when creating a peer, but only sometimes they actually open. However, when adding video/audio later, the datachannels open! So something strange is happening there with the offer/answering with only datachannels.
Also, the 'peer.ondatachannel' object isn't correctly attached to unity yet, so only the caller can send data over the channel but the receiver is not pointing to any datachannel so cannot send.

**Comment by antoined73** [comment]
Nice ! 
Could you make another PR, so that this repository keep synced with yours ?

I have played with the official 2.4.0-exp3 lib, connecting via WebRTC with a custom signaling server. The connection is made, the sdp and ice candidates are both exchanged, but then when it is time to receive the video in unity, the texture becomes greyish, and he video does not appear. I don't know what happens, because it is working in Android, but not in Windows build nor Windows Unity Editor. So maybe it is the same problem you encountered ? If so, then the official library is already having this sorts of problems, I'll try your new update and play with WebGL also, and try to send the video later.

**Comment by rollersteaam** [comment]
@antoined73 No idea about how this repository may have changed it, but `com.unity.webrtc` expects its native plugin to be located in `Packages/com.unity.webrtc/Runtime/Plugins/x86_64/webrtc.dll` and not `**Assets**/Packages/com.unity.webrtc...` Having the package in Assets may be causing issues. You can change the path to the dll in the source. This may entirely not fix your issue however

**Comment by antoined73** [comment]
@rollersteaam Yes `com.unity.webrtc` expect its native pugin to be located in Packages because the path is written in WebRTC.cs :
```
public static class WebRTC
    {
#if UNITY_EDITOR_OSX
        internal const string Lib = "Packages/com.unity.webrtc/Runtime/Plugins/x86_64/webrtc.bundle/Contents/MacOS/webrtc";
#elif UNITY_EDITOR_LINUX
        internal const string Lib = "Packages/com.unity.webrtc/Runtime/Plugins/x86_64/libwebrtc.so";
#elif UNITY_EDITOR_WIN
        internal const string Lib = "Assets/com.unity.webrtc/Runtime/Plugins/x86_64/webrtc.dll"; // "Packages/com.unity.webrtc/Runtime/Plugins/x86_64/webrtc.dll";
#elif UNITY_STANDALONE
        internal const string Lib = "webrtc";
#elif UNITY_IOS || UNITY_WEBGL
        internal const string Lib = "__Internal";
#elif UNITY_ANDROID
        internal const string Lib = "webrtc";
#endif
```

But as you can see, by changing the location like so, the lib is loading. So I don't think that this is the issue there
